include/sparrow/nullable.hpp

```diff
@@ -0,0 +1,659 @@
+// Copyright 2024 Man Group Operations Limited
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <compare>
+#include <concepts>
+#include <exception>
+#include <type_traits>
+
+#include "sparrow/mp_utils.hpp"
+
+namespace sparrow
+{
+    using mpl::boolean_like;
+
+    template <class T, boolean_like B>
+    class nullable;
+    /*
+     * Default traits for the nullable class. These traits should be specialized
+     * for proxy classes whose reference and const_reference types are not
+     * defined as usual. For instance:
+     *
+     * @code{.cpp}
+     * struct nullable_traits<string_proxy>
+     * {
+     *     using value_type = std::string;
+     *     using reference = string_proxy;
+     *     using const_reference = std::string_view;
+     *     using rvalue_reverence = std::string&&;
+     *     using const_rvalue_reference = const std::string&&;
+     * };
+     * @endcode
+     */
+    template <class T>
+    struct nullable_traits
+    {
+        using value_type = T;
+        using reference = std::add_lvalue_reference_t<value_type>;
+        using const_reference = std::add_lvalue_reference_t<std::add_const_t<value_type>>;
+        using rvalue_reference = value_type&&;
+        using const_rvalue_reference = const value_type&&;
+    };
+
+    template <class T>
+    struct nullable_traits<T&>
+    {
+        using value_type = T;
+        using reference = std::add_lvalue_reference_t<value_type>;
+        using const_reference = std::add_lvalue_reference_t<std::add_const_t<value_type>>;
+        using rvalue_reference = reference;
+        using const_rvalue_reference = const_reference;
+    };
+
+    /*
+     * Defines a type of object to be thrown by nullable::value when accessing
+     * a nullable object whose value is null.
+     */
+    class bad_nullable_access : public std::exception
+    {
+    public:
+
+        bad_nullable_access() noexcept = default;
+        bad_nullable_access(const bad_nullable_access&) noexcept = default;
+        bad_nullable_access& operator=(const bad_nullable_access&) noexcept = default;
+
+        virtual const char* what() const noexcept
+        {
+            return message;
+        }
+
+    private:
+
+        static constexpr const char* message = "Invalid access to null value";
```

Comment:
```suggestion
        static constexpr const char* message = "Invalid access to null arrow value";
```

This idea of null value (which does not destroy the object) is specific to arrow and should be clear in the error, otherwise one would look for null pointers etc.

Reply:
It's more specific to libraries where you can have missing / null values. We could imagine moving this class to a util library later, taht could be used by data frame libraries / applications.